 <img src="https://github.com/MahendraCherukupalli/The-Sparks-Foundation-Internship/blob/master/wpl_badge1.png?raw=true"/>

## Steps:
- Import Libraries
- Get the Data
- Data Preprocessing
- Train-Test Split
- Scalling the Data
- Creating and Training the Model
- Predictions
- Residual Analysis
- Evaluating the modelm

**Let's get started!**

# Import Libraries

In [1]:
import warnings
warnings.filterwarnings('ignore')

In [2]:
import numpy as np
import pandas as pd
import seaborn as sns  
import matplotlib.pyplot as plt  
%matplotlib inline

# Get the data

In [3]:
data=pd.read_csv("matches.csv")
data.head()

id  season       city        date                        team1  \
0   1    2017  Hyderabad  2017-04-05          Sunrisers Hyderabad   
1   2    2017       Pune  2017-04-06               Mumbai Indians   
2   3    2017     Rajkot  2017-04-07                Gujarat Lions   
3   4    2017     Indore  2017-04-08       Rising Pune Supergiant   
4   5    2017  Bangalore  2017-04-08  Royal Challengers Bangalore   

                         team2                  toss_winner toss_decision  \
0  Royal Challengers Bangalore  Royal Challengers Bangalore         field   
1       Rising Pune Supergiant       Rising Pune Supergiant         field   
2        Kolkata Knight Riders        Kolkata Knight Riders         field   
3              Kings XI Punjab              Kings XI Punjab         field   
4             Delhi Daredevils  Royal Challengers Bangalore           bat   

   result  dl_applied                       winner  win_by_runs  \
0  normal           0          Sunrisers Hyderabad           35   
1  normal           0       Rising Pune Supergiant            0   
2  normal           0        Kolkata Knight Riders            0   
3  normal           0              Kings XI Punjab            0   
4  normal           0  Royal Challengers Bangalore           15   

   win_by_wickets player_of_match                                      venue  \
0               0    Yuvraj Singh  Rajiv Gandhi International Stadium, Uppal   
1               7       SPD Smith    Maharashtra Cricket Association Stadium   
2              10         CA Lynn     Saurashtra Cricket Association Stadium   
3               6      GJ Maxwell                     Holkar Cricket Stadium   
4               0       KM Jadhav                      M Chinnaswamy Stadium   

          umpire1        umpire2 umpire3  
0     AY Dandekar       NJ Llong     NaN  
1  A Nand Kishore         S Ravi     NaN  
2     Nitin Menon      CK Nandan     NaN  
3    AK Chaudhary  C Shamshuddin     NaN  
4             NaN            NaN     NaN

In [4]:
data.shape

(756, 18)

In [5]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 756 entries, 0 to 755
Data columns (total 18 columns):
id                 756 non-null int64
season             756 non-null int64
city               749 non-null object
date               756 non-null object
team1              756 non-null object
team2              756 non-null object
toss_winner        756 non-null object
toss_decision      756 non-null object
result             756 non-null object
dl_applied         756 non-null int64
winner             752 non-null object
win_by_runs        756 non-null int64
win_by_wickets     756 non-null int64
player_of_match    752 non-null object
venue              756 non-null object
umpire1            754 non-null object
umpire2            754 non-null object
umpire3            119 non-null object
dtypes: int64(5), object(13)
memory usage: 106.4+ KB


In [6]:
data.describe()

id       season  dl_applied  win_by_runs  win_by_wickets
count    756.000000   756.000000  756.000000   756.000000      756.000000
mean    1792.178571  2013.444444    0.025132    13.283069        3.350529
std     3464.478148     3.366895    0.156630    23.471144        3.387963
min        1.000000  2008.000000    0.000000     0.000000        0.000000
25%      189.750000  2011.000000    0.000000     0.000000        0.000000
50%      378.500000  2013.000000    0.000000     0.000000        4.000000
75%      567.250000  2016.000000    0.000000    19.000000        6.000000
max    11415.000000  2019.000000    1.000000   146.000000       10.000000

In [7]:
data1=pd.read_csv("deliveries.csv")
data1.head()

match_id  inning         batting_team                 bowling_team  over  \
0         1       1  Sunrisers Hyderabad  Royal Challengers Bangalore     1   
1         1       1  Sunrisers Hyderabad  Royal Challengers Bangalore     1   
2         1       1  Sunrisers Hyderabad  Royal Challengers Bangalore     1   
3         1       1  Sunrisers Hyderabad  Royal Challengers Bangalore     1   
4         1       1  Sunrisers Hyderabad  Royal Challengers Bangalore     1   

   ball    batsman non_striker    bowler  is_super_over  ...  bye_runs  \
0     1  DA Warner    S Dhawan  TS Mills              0  ...         0   
1     2  DA Warner    S Dhawan  TS Mills              0  ...         0   
2     3  DA Warner    S Dhawan  TS Mills              0  ...         0   
3     4  DA Warner    S Dhawan  TS Mills              0  ...         0   
4     5  DA Warner    S Dhawan  TS Mills              0  ...         0   

   legbye_runs  noball_runs  penalty_runs  batsman_runs  extra_runs  \
0            0            0             0             0           0   
1            0            0             0             0           0   
2            0            0             0             4           0   
3            0            0             0             0           0   
4            0            0             0             0           2   

   total_runs  player_dismissed dismissal_kind fielder  
0           0               NaN            NaN     NaN  
1           0               NaN            NaN     NaN  
2           4               NaN            NaN     NaN  
3           0               NaN            NaN     NaN  
4           2               NaN            NaN     NaN  

[5 rows x 21 columns]

In [8]:
data1.shape

(179078, 21)

In [9]:
data1.describe()

match_id         inning           over           ball  \
count  179078.000000  179078.000000  179078.000000  179078.000000   
mean     1802.252957       1.482952      10.162488       3.615587   
std      3472.322805       0.502074       5.677684       1.806966   
min         1.000000       1.000000       1.000000       1.000000   
25%       190.000000       1.000000       5.000000       2.000000   
50%       379.000000       1.000000      10.000000       4.000000   
75%       567.000000       2.000000      15.000000       5.000000   
max     11415.000000       5.000000      20.000000       9.000000   

       is_super_over      wide_runs       bye_runs    legbye_runs  \
count  179078.000000  179078.000000  179078.000000  179078.000000   
mean        0.000452       0.036721       0.004936       0.021136   
std         0.021263       0.251161       0.116480       0.194908   
min         0.000000       0.000000       0.000000       0.000000   
25%         0.000000       0.000000       0.000000       0.000000   
50%         0.000000       0.000000       0.000000       0.000000   
75%         0.000000       0.000000       0.000000       0.000000   
max         1.000000       5.000000       4.000000       5.000000   

         noball_runs   penalty_runs   batsman_runs     extra_runs  \
count  179078.000000  179078.000000  179078.000000  179078.000000   
mean        0.004183       0.000056       1.246864       0.067032   
std         0.070492       0.016709       1.608270       0.342553   
min         0.000000       0.000000       0.000000       0.000000   
25%         0.000000       0.000000       0.000000       0.000000   
50%         0.000000       0.000000       1.000000       0.000000   
75%         0.000000       0.000000       1.000000       0.000000   
max         5.000000       5.000000       7.000000       7.000000   

          total_runs  
count  179078.000000  
mean        1.313897  
std         1.605422  
min         0.000000  
25%         0.000000  
50%         1.000000  
75%         1.000000  
max        10.000000

# Data Preprocessing

In [10]:
# checking for null values...
data.isnull().sum()

id                   0
season               0
city                 7
date                 0
team1                0
team2                0
toss_winner          0
toss_decision        0
result               0
dl_applied           0
winner               4
win_by_runs          0
win_by_wickets       0
player_of_match      4
venue                0
umpire1              2
umpire2              2
umpire3            637
dtype: int64

In [11]:
data1.isnull().sum()

match_id                 0
inning                   0
batting_team             0
bowling_team             0
over                     0
ball                     0
batsman                  0
non_striker              0
bowler                   0
is_super_over            0
wide_runs                0
bye_runs                 0
legbye_runs              0
noball_runs              0
penalty_runs             0
batsman_runs             0
extra_runs               0
total_runs               0
player_dismissed    170244
dismissal_kind      170244
fielder             172630
dtype: int64

In [12]:
data.head(10)

id  season       city        date                        team1  \
0   1    2017  Hyderabad  2017-04-05          Sunrisers Hyderabad   
1   2    2017       Pune  2017-04-06               Mumbai Indians   
2   3    2017     Rajkot  2017-04-07                Gujarat Lions   
3   4    2017     Indore  2017-04-08       Rising Pune Supergiant   
4   5    2017  Bangalore  2017-04-08  Royal Challengers Bangalore   
5   6    2017  Hyderabad  2017-04-09                Gujarat Lions   
6   7    2017     Mumbai  2017-04-09        Kolkata Knight Riders   
7   8    2017     Indore  2017-04-10  Royal Challengers Bangalore   
8   9    2017       Pune  2017-04-11             Delhi Daredevils   
9  10    2017     Mumbai  2017-04-12          Sunrisers Hyderabad   

                         team2                  toss_winner toss_decision  \
0  Royal Challengers Bangalore  Royal Challengers Bangalore         field   
1       Rising Pune Supergiant       Rising Pune Supergiant         field   
2        Kolkata Knight Riders        Kolkata Knight Riders         field   
3              Kings XI Punjab              Kings XI Punjab         field   
4             Delhi Daredevils  Royal Challengers Bangalore           bat   
5          Sunrisers Hyderabad          Sunrisers Hyderabad         field   
6               Mumbai Indians               Mumbai Indians         field   
7              Kings XI Punjab  Royal Challengers Bangalore           bat   
8       Rising Pune Supergiant       Rising Pune Supergiant         field   
9               Mumbai Indians               Mumbai Indians         field   

   result  dl_applied                       winner  win_by_runs  \
0  normal           0          Sunrisers Hyderabad           35   
1  normal           0       Rising Pune Supergiant            0   
2  normal           0        Kolkata Knight Riders            0   
3  normal           0              Kings XI Punjab            0   
4  normal           0  Royal Challengers Bangalore           15   
5  normal           0          Sunrisers Hyderabad            0   
6  normal           0               Mumbai Indians            0   
7  normal           0              Kings XI Punjab            0   
8  normal           0             Delhi Daredevils           97   
9  normal           0               Mumbai Indians            0   

   win_by_wickets player_of_match                                      venue  \
0               0    Yuvraj Singh  Rajiv Gandhi International Stadium, Uppal   
1               7       SPD Smith    Maharashtra Cricket Association Stadium   
2              10         CA Lynn     Saurashtra Cricket Association Stadium   
3               6      GJ Maxwell                     Holkar Cricket Stadium   
4               0       KM Jadhav                      M Chinnaswamy Stadium   
5               9     Rashid Khan  Rajiv Gandhi International Stadium, Uppal   
6               4          N Rana                           Wankhede Stadium   
7               8        AR Patel                     Holkar Cricket Stadium   
8               0       SV Samson    Maharashtra Cricket Association Stadium   
9               4       JJ Bumrah                           Wankhede Stadium   

          umpire1        umpire2 umpire3  
0     AY Dandekar       NJ Llong     NaN  
1  A Nand Kishore         S Ravi     NaN  
2     Nitin Menon      CK Nandan     NaN  
3    AK Chaudhary  C Shamshuddin     NaN  
4             NaN            NaN     NaN  
5      A Deshmukh       NJ Llong     NaN  
6     Nitin Menon      CK Nandan     NaN  
7    AK Chaudhary  C Shamshuddin     NaN  
8     AY Dandekar         S Ravi     NaN  
9     Nitin Menon      CK Nandan     NaN

In [13]:
SRH=data["team1"]=='Sunrisers Hyderabad'
Dc=data["team1"]=='Delhi Daredevils'
SRH1=data["team2"]=='Sunrisers Hyderabad'
Dc1=data["team2"]=='Delhi Daredevils'

In [14]:
rdf=data[(SRH & Dc1)|(SRH1 & Dc)]

In [15]:
rdf.head()

id  season       city        date                team1  \
20    21    2017  Hyderabad  2017-04-19  Sunrisers Hyderabad   
38    39    2017      Delhi  2017-05-02  Sunrisers Hyderabad   
393  394    2013      Delhi  2013-04-12     Delhi Daredevils   
427  428    2013  Hyderabad  2013-05-04     Delhi Daredevils   
468  469    2014        NaN  2014-04-25  Sunrisers Hyderabad   

                   team2          toss_winner toss_decision  result  \
20      Delhi Daredevils  Sunrisers Hyderabad           bat  normal   
38      Delhi Daredevils     Delhi Daredevils         field  normal   
393  Sunrisers Hyderabad     Delhi Daredevils           bat  normal   
427  Sunrisers Hyderabad     Delhi Daredevils           bat  normal   
468     Delhi Daredevils  Sunrisers Hyderabad           bat  normal   

     dl_applied               winner  win_by_runs  win_by_wickets  \
20            0  Sunrisers Hyderabad           15               0   
38            0     Delhi Daredevils            0               6   
393           0  Sunrisers Hyderabad            0               3   
427           0  Sunrisers Hyderabad            0               6   
468           0  Sunrisers Hyderabad            4               0   

    player_of_match                                      venue      umpire1  \
20    KS Williamson  Rajiv Gandhi International Stadium, Uppal  CB Gaffaney   
38   Mohammed Shami                           Feroz Shah Kotla     YC Barde   
393        A Mishra                           Feroz Shah Kotla    Aleem Dar   
427       DJG Sammy  Rajiv Gandhi International Stadium, Uppal    Asad Rauf   
468        AJ Finch        Dubai International Cricket Stadium    M Erasmus   

         umpire2 umpire3  
20      NJ Llong     NaN  
38   Nitin Menon     NaN  
393  Subroto Das     NaN  
427     S Asnani     NaN  
468       S Ravi     NaN

In [16]:
rdf.shape

(12, 18)

In [17]:
rdf[["team1","team2","toss_winner","toss_decision","winner"]]

team1                team2          toss_winner  \
20   Sunrisers Hyderabad     Delhi Daredevils  Sunrisers Hyderabad   
38   Sunrisers Hyderabad     Delhi Daredevils     Delhi Daredevils   
393     Delhi Daredevils  Sunrisers Hyderabad     Delhi Daredevils   
427     Delhi Daredevils  Sunrisers Hyderabad     Delhi Daredevils   
468  Sunrisers Hyderabad     Delhi Daredevils  Sunrisers Hyderabad   
488     Delhi Daredevils  Sunrisers Hyderabad  Sunrisers Hyderabad   
530     Delhi Daredevils  Sunrisers Hyderabad     Delhi Daredevils   
560  Sunrisers Hyderabad     Delhi Daredevils  Sunrisers Hyderabad   
617  Sunrisers Hyderabad     Delhi Daredevils     Delhi Daredevils   
627  Sunrisers Hyderabad     Delhi Daredevils     Delhi Daredevils   
671     Delhi Daredevils  Sunrisers Hyderabad     Delhi Daredevils   
677     Delhi Daredevils  Sunrisers Hyderabad     Delhi Daredevils   

    toss_decision               winner  
20            bat  Sunrisers Hyderabad  
38          field     Delhi Daredevils  
393           bat  Sunrisers Hyderabad  
427           bat  Sunrisers Hyderabad  
468           bat  Sunrisers Hyderabad  
488         field  Sunrisers Hyderabad  
530           bat     Delhi Daredevils  
560           bat  Sunrisers Hyderabad  
617         field     Delhi Daredevils  
627         field     Delhi Daredevils  
671           bat  Sunrisers Hyderabad  
677           bat  Sunrisers Hyderabad

In [18]:
rdf[["team1","team2","toss_winner","toss_decision","winner"]].isnull().sum()

team1            0
team2            0
toss_winner      0
toss_decision    0
winner           0
dtype: int64

# 1

In [44]:
rdf[(rdf["toss_winner"].isin(["Delhi Daredevils","Delhi Capitals"]))&(rdf["toss_decision"]=="field")]

id  season       city        date                team1  \
38    39    2017      Delhi  2017-05-02  Sunrisers Hyderabad   
617  618    2016  Hyderabad  2016-05-12  Sunrisers Hyderabad   
627  628    2016     Raipur  2016-05-20  Sunrisers Hyderabad   

                team2       toss_winner toss_decision  result  dl_applied  \
38   Delhi Daredevils  Delhi Daredevils         field  normal           0   
617  Delhi Daredevils  Delhi Daredevils         field  normal           0   
627  Delhi Daredevils  Delhi Daredevils         field  normal           0   

               winner  win_by_runs  win_by_wickets player_of_match  \
38   Delhi Daredevils            0               6  Mohammed Shami   
617  Delhi Daredevils            0               7       CH Morris   
627  Delhi Daredevils            0               6         KK Nair   

                                                venue         umpire1  \
38                                   Feroz Shah Kotla        YC Barde   
617         Rajiv Gandhi International Stadium, Uppal      K Bharatan   
627  Shaheed Veer Narayan Singh International Stadium  A Nand Kishore   

          umpire2 umpire3  
38    Nitin Menon     NaN  
617     M Erasmus     NaN  
627  BNJ Oxenford     NaN

In [45]:
rdf["winner"].value_counts()

Sunrisers Hyderabad    8
Delhi Daredevils       4
Name: winner, dtype: int64

according to the dataset SRH win the match but DC performs well in 2020

# 2

In [21]:
Total_runs=data1[(data1["batting_team"].isin(["Sunrisers Hyderabad","Delhi Daredevils","Delhi Capitals"]))&(data1["bowling_team"].isin(["Sunrisers Hyderabad","Delhi Daredevils","Delhi Capitals"]))]

In [22]:
Total_runs.groupby(["match_id"])["total_runs"].sum()

match_id
21       367
39       374
394      229
428      161
469      364
489      187
531      330
561      320
618      296
628      319
7929     336
7935     383
11152    276
11321    285
11413    341
Name: total_runs, dtype: int64

In [23]:
Total_runs.groupby(["match_id"])["total_runs"].sum().mean()

304.53333333333336

according to the dataset both teams score average of 305 runs 

# 3

In [24]:
wickets=data1[(data1["batting_team"].isin(["Sunrisers Hyderabad","Delhi Daredevils","Delhi Capitals"]))&(data1["bowling_team"].isin(["Sunrisers Hyderabad","Delhi Daredevils","Delhi Capitals"]))]

In [25]:
wickets1=wickets[wickets["dismissal_kind"].isin(["caught"])]

In [26]:
wickets1.groupby(["match_id"])["dismissal_kind"].count()

match_id
21        7
39        5
394      10
428       4
469       5
489       7
531       6
561       5
618       5
628       6
7929      3
7935      3
11152    10
11321    13
11413    11
Name: dismissal_kind, dtype: int64

In [27]:
wickets1.groupby(["match_id"])["dismissal_kind"].count().mean()

6.666666666666667

According to dataset both teams loss average of 7 wickets by caught

In [28]:
data.columns

Index(['id', 'season', 'city', 'date', 'team1', 'team2', 'toss_winner',
       'toss_decision', 'result', 'dl_applied', 'winner', 'win_by_runs',
       'win_by_wickets', 'player_of_match', 'venue', 'umpire1', 'umpire2',
       'umpire3'],
      dtype='object')

In [29]:
data1.columns

Index(['match_id', 'inning', 'batting_team', 'bowling_team', 'over', 'ball',
       'batsman', 'non_striker', 'bowler', 'is_super_over', 'wide_runs',
       'bye_runs', 'legbye_runs', 'noball_runs', 'penalty_runs',
       'batsman_runs', 'extra_runs', 'total_runs', 'player_dismissed',
       'dismissal_kind', 'fielder'],
      dtype='object')

# 4

In [30]:
Dc_6=data1[(data1["batting_team"].isin(["Delhi Daredevils","Delhi Capitals"]))&(data1["bowling_team"]=="Sunrisers Hyderabad")]

In [31]:
Dc_6n=Dc_6[Dc_6["over"].isin([1,2,3,4,5,6])]

In [32]:
Dc_6n.groupby(["match_id"])["total_runs"].sum()

match_id
21       56
39       62
394      29
428      26
469      44
489      52
531      40
561      49
618      50
628      37
7929     61
7935     39
11152    37
11321    53
11413    56
Name: total_runs, dtype: int64

In [33]:
Dc_6n.groupby(["match_id"])["total_runs"].sum().mean()

46.06666666666667

In [34]:
SRH_6=data1[(data1["batting_team"]=="Sunrisers Hyderabad")&(data1["bowling_team"].isin(["Delhi Daredevils","Delhi Capitals"]))]

In [35]:
SRH_6n=SRH_6[SRH_6["over"].isin([1,2,3,4,5,6])]

In [36]:
SRH_6n.groupby(["match_id"])["total_runs"].sum()

match_id
21       39
39       66
394      38
428      41
469      55
489      44
531      50
561      41
618      51
628      46
7929     64
7935     52
11152    71
11321    41
11413    54
Name: total_runs, dtype: int64

In [37]:
SRH_6n.groupby(["match_id"])["total_runs"].sum().mean()

50.2

According to dataset Dc score 4 runs less than SRH

# 5

In [38]:
extras=data1[(data1["batting_team"].isin(["Sunrisers Hyderabad","Delhi Daredevils","Delhi Capitals"]))&(data1["bowling_team"].isin(["Sunrisers Hyderabad","Delhi Daredevils","Delhi Capitals"]))]

In [39]:
df=pd.DataFrame(extras.groupby(["match_id"])["extra_runs",'wide_runs','bye_runs', 'legbye_runs', 'noball_runs', 'penalty_runs'].sum())

In [40]:
df["total_extras"]=df["extra_runs"]+df["wide_runs"]+df["bye_runs"]+df["legbye_runs"]+df["noball_runs"]+df["penalty_runs"]

In [41]:
df

extra_runs  wide_runs  bye_runs  legbye_runs  noball_runs  \
match_id                                                              
21                11          3         3            3            2   
39                11          4         0            6            1   
394               10          5         0            3            2   
428                9          6         1            1            1   
469               11          4         0            6            1   
489               16          5         4            6            1   
531               13          7         3            3            0   
561                8          4         0            4            0   
618               14          6         0            6            2   
628               17         11         1            5            0   
7929               9          7         0            1            1   
7935               5          2         0            3            0   
11152             16          5         4            7            0   
11321             14          7         0            5            2   
11413             14         10         1            2            1   

          penalty_runs  total_extras  
match_id                              
21                   0            22  
39                   0            22  
394                  0            20  
428                  0            18  
469                  0            22  
489                  0            32  
531                  0            26  
561                  0            16  
618                  0            28  
628                  0            34  
7929                 0            18  
7935                 0            10  
11152                0            32  
11321                0            28  
11413                0            28

In [42]:
df["total_extras"].mean()

23.733333333333334

According to dataset both team give average of 24 extra runs